src/nitypes/waveform/_digital/_waveform.py

```diff
@@ -66,6 +66,9 @@ class DigitalWaveformFailure:
     signal_index: int
     """The signal index where the test failure occurred."""
 
+    data_index: int
```

Comment:
Adding a field to the middle of a dataclass is a breaking change because it adds a required positional parameter to its constructor. However, you can argue that clients are unlikely to construct failure objects outside of tests, and that there are probably not any clients doing this yet.

I think we should also consider whether adding the data index to the failure object is necessary. Given a failure, you can get the data index with `waveform.signals[failure.signal_index].data_index`.